feat(assets): document more utilities (#11194)
withastro/docs: src/content/docs/en/reference/image-service-reference.mdx

const isProtocolMatch3 = matchProtocol(testURL1);
console.log(`Does the protocol match (no protocol specified)? ${isProtocolMatch3}`); // Output: true
```

### `isESMImportedImage()`

<p>
  **Type:** `(src: ImageMetadata | string): boolean`<br />
  <Since v="4.0.0" />
</p>

Determines if the given source is an ECMAScript Module (ESM) imported image.

```ts

> import {isESMImportedImage} from 'astro/assets/utils';

Reviewer: To be consistent with the formatting of other imports in the page:
```suggestion
import { isESMImportedImage } from 'astro/assets/utils';
```

// Example usage of isESMImportedImage
const imageMetadataExample = {
  src: '/images/photo.jpg',
  width: 800,
  height: 600,
  format: 'jpg',
};

const filePathExample = '/images/photo.jpg';

// Check if the input is an ESM imported image
const isMetadataImage = isESMImportedImage(imageMetadataExample);
console.log(`Is imageMetadataExample an ESM imported image? ${isMetadataImage}`); // Output: true

const isFilePathImage = isESMImportedImage(filePathExample);
console.log(`Is filePathExample an ESM imported image? ${isFilePathImage}`); // Output: false
```

### `isRemoteImage()`

<p>
  **Type:** `(src: ImageMetadata | string): boolean`<br />
  <Since v="4.0.0" />
</p>

Determines if the provided source is a remote image URL in the form of a string.

```ts

> import {isRemoteImage} from 'astro/assets/utils';

Reviewer: Same here, just formatting:
```suggestion
import { isRemoteImage } from 'astro/assets/utils';
```

  <Since v="4.0.0" />
</p>

Extracts image metadata such as dimensions, format, and orientation from the provided image data.

```ts
import { imageMetadata } from 'astro/assets/utils';

async function extractImageMetadata() {
  // Example image data (Uint8Array)
  const exampleImageData = new Uint8Array([/* ...binary image data... */]);

  // Optional source path (useful for debugging or additional metadata context)
  const sourcePath = '/images/photo.jpg';

  try {
    // Extract metadata from the image data
    const metadata = await imageMetadata(exampleImageData, sourcePath);

    console.log('Extracted Image Metadata:', metadata);
    // Example output:
    // {
    //   width: 800,
    //   height: 600,
    //   format: 'jpg',
    //   orientation: undefined
    // }
  } catch (error) {
    console.error('Failed to extract metadata from image:', error);
  }
}

await extractImageMetadata();
```

### `emitESMImage()`

<p>
  **Type:** `(id: string | undefined, _watchMode: boolean, experimentalSvgEnabled: boolean,  fileEmitter?: FileEmitter): Promise<ImageMetadataWithContents | undefined>`<br />
  <Since v="4.0.0" />
</p>

> Processes an image file and emits its metadata and optionally its contents. In build mode, the function uses `fileEmitter` to generate an asset reference. In development mode, it resolves to a local file URL with query parameters for metadata.

Author: Yes, you got everything right. The rest is an implementation detail.

IMHO we should create a brand new function and deprecate this one. It's unfortunate that we exposed it, forgot to documented it, and now it's a bit broken 😅

const isProtocolMatch3 = matchProtocol(testURL1);
console.log(`Does the protocol match (no protocol specified)? ${isProtocolMatch3}`); // Output: true
```

> ### `isESMImportedImage`

Reviewer: To be consistent since this is also a function:
```suggestion
### `isESMImportedImage()`
```

<p>

>   **Type:** `(src: ImageMetadata | string): src is ImageMetadata`<br />

Reviewer: I think for the docs we can simplify a bit the type for those less familiar with type predicates:
```suggestion
  **Type:** `(src: ImageMetadata | string): boolean`<br />
```

  <Since v="4.0.0" />
</p>

Determines if the given source is an ECMAScript Module (ESM) imported image.

```ts
import {isESMImportedImage} from 'astro/assets/utils';

// Example usage of isESMImportedImage
const imageMetadataExample = {
  src: '/images/photo.jpg',
  width: 800,
  height: 600,
  format: 'jpg',
};

const filePathExample = '/images/photo.jpg';

// Check if the input is an ESM imported image
const isMetadataImage = isESMImportedImage(imageMetadataExample);
console.log(`Is imageMetadataExample an ESM imported image? ${isMetadataImage}`); // Output: true

const isFilePathImage = isESMImportedImage(filePathExample);
console.log(`Is filePathExample an ESM imported image? ${isFilePathImage}`); // Output: false
```

### `isRemoteImage()`

<p>

>   **Type:** `(src: ImageMetadata | string): src is string`<br />

Reviewer: Same as above, we can simplify the type a bit:
```suggestion
  **Type:** `(src: ImageMetadata | string): boolean`<br />
```

  <Since v="4.0.0" />
</p>

Determines if the provided source is a remote image URL in the form of a string.

```ts
import {isRemoteImage} from 'astro/assets/utils';

// Example usage of isRemoteImage
const remoteImageUrl = 'https://example.com/images/photo.jpg';
const localImageMetadata = {
  src: '/images/photo.jpg',
  width: 800,
  height: 600,
  format: 'jpg',
};

// Check if the input is a remote image URL
const isRemote1 = isRemoteImage(remoteImageUrl);
console.log(`Is remoteImageUrl a remote image? ${isRemote1}`); // Output: true

const isRemote2 = isRemoteImage(localImageMetadata);
console.log(`Is localImageMetadata a remote image? ${isRemote2}`); // Output: false
```

### `resolveSrc()`

<p>
  **Type:** `(src: UnresolvedImageTransform['src']): Promise<string | ImageMetadata>`<br />
  <Since v="4.0.0" />
</p>

> Resolves the source of an image transformation by handling asynchronous or synchronous inputs.

Reviewer: I feel like we should detail a bit more this one for the docs. By reading the source code it seems this is only really helpful if we pass an imported image as a Promise. So this is useful when we don't know if the input type.

I have to admit right now I'm having a bit of trouble finding the right wording... Perhaps something like:
```suggestion
Returns the image source. This function ensures that if `src` is a Promise (e.g., a dynamic `import()`), it is awaited and the correct `src` is extracted. If `src` is already a resolved value, it is returned as-is.
```

> ```ts
> import {resolveSrc} from 'astro/assets/utils';
> 
> // Example usage of resolveSrc with a local image
> const localImage = {
>   src: '/images/photo.jpg',
>   width: 800,
>   height: 600,
>   format: 'jpg',
> };
> 
> const remoteImage = 'https://example.com/images/photo.jpg';
> 
> async function exampleResolveSrc() {
>   // Resolve a local image
>   const resolvedLocal = await resolveSrc(localImage);
>   console.log(`Resolved local image source:`, resolvedLocal);
>   
>   // Resolve a remote image URL
>   const resolvedRemote = await resolveSrc(remoteImage);
>   console.log(`Resolved remote image source:`, resolvedRemote);
>   }
> 
> await exampleResolveSrc();
> ```

Reviewer: I think we don't need the wrapper function here. But I also feel like the example is not really helpful... because in both cases, if I understood correctly the function, it will return the exact input. Perhaps we could show the three cases:
````suggestion
```ts
import {resolveSrc} from 'astro/assets/utils';
import localImage from "./images/photo.jpg";

const resolvedLocal = await resolveSrc(localImage);
// will be `{ src: '/images/photo.jpg', width: 800, height: 600, format: 'jpg' }`

const resolvedRemote = await resolveSrc("https://example.com/remote-img.jpg");
// will be `"https://example.com/remote-img.jpg"`

const resolvedDynamic = await resolveSrc(import("./images/dynamic-image.jpg"))
// will be `{ src: '/images/dynamic-image.jpg', width: 800, height: 600, format: 'jpg' }`
```
````

### `imageMetadata()`

<p>

>   **Type:** `(data: Uint8Array, src?: string): Promise<Omit<ImageMetadata, 'src' | 'fsPath'>`<br />

Reviewer: typo:
```suggestion
  **Type:** `(data: Uint8Array, src?: string): Promise<Omit<ImageMetadata, 'src' | 'fsPath'>>`<br />
```

  <Since v="4.0.0" />
</p>

Extracts image metadata such as dimensions, format, and orientation from the provided image data.

```ts
import { imageMetadata } from 'astro/assets/utils';

async function extractImageMetadata() {
  // Example image data (Uint8Array)
  const exampleImageData = new Uint8Array([/* ...binary image data... */]);

  // Optional source path (useful for debugging or additional metadata context)
  const sourcePath = '/images/photo.jpg';

  try {
    // Extract metadata from the image data
    const metadata = await imageMetadata(exampleImageData, sourcePath);

    console.log('Extracted Image Metadata:', metadata);
    // Example output:
    // {
    //   width: 800,
    //   height: 600,
    //   format: 'jpg',
    //   orientation: undefined
    // }
  } catch (error) {
    console.error('Failed to extract metadata from image:', error);
  }
}

await extractImageMetadata();
```

### `emitESMImage()`

<p>
  **Type:** `(id: string | undefined, _watchMode: boolean, experimentalSvgEnabled: boolean,  fileEmitter?: FileEmitter): Promise<ImageMetadataWithContents | undefined>`<br />
  <Since v="4.0.0" />
</p>

> Processes an image file and emits its metadata and optionally its contents. This function supports both build and development modes.

Reviewer: Looking at the implementation, I feel like an important difference between build and dev is the generated `src` so perhaps we should explain this:
```suggestion
Processes an image file and emits its metadata and optionally its contents. In build mode, the function uses `fileEmitter` to generate an asset reference. In development mode, it resolves to a local file URL with query parameters for metadata.
```

```ts

import { emitESMImage } from 'astro/assets/utils';

async function processImage() {
  const imageId = '/images/photo.jpg';
  // DEPRECATED!!
  const watchMode = false; // Set to true in development mode
  const experimentalSvgEnabled = true; // Experimental SVG support
  
  try {
    const result = await emitESMImage(imageId, watchMode, experimentalSvgEnabled);
    if (result) {
      console.log('Image metadata with contents:', result);
      // Example output:
      // {
      //   width: 800,
      //   height: 600,
      //   format: 'jpg',
      //   contents: Uint8Array([...])
      // }
    } else {
      console.log('No metadata was emitted for this image.');
    }
  } catch (error) {
    console.error('Failed to emit ESM image:', error);
  }
}

await processImage();

> ```

Author: I pushed the changes is [`ceaca7c` (#11194)](https://github.com/withastro/docs/pull/11194/commits/ceaca7ccfdf09df121910de5412388ff7ac2d27a)

### `getOrigQueryParams()`

<p>
  **Type:** `(params: URLSearchParams): Pick<ImageMetadata, 'width' | 'height' | 'format'> | undefined>`<br />
  <Since v="4.0.0" />
</p>

Extracts the original image query parameters (width, height, format) from the given `URLSearchParams` object and returns them as an object. If any of the required parameters are missing or invalid, the function returns undefined.
 

> The `width` and `height` are parsed to integer values.

Author: I belive it's an information that we can remove, since it can be inferred from the returned type :)
